File: 0509-fibonacci-number/0509-fibonacci-number.py
class Solution:
    def fib(self, n: int) -> int:
        dp = [-1] * (n + 1)

        def fn(i):
            if i <= 1:
                return i

            if dp[i] != -1:
                return dp[i]

            dp[i] = fn(i - 1) + fn(i - 2)
            return dp[i]

        return fn(n)

File: 0509-fibonacci-number/0509-fibonacci-number.java
class Solution {
    static HashMap<Integer,Integer> dp =new HashMap<>();
    public static int fn(int n)
    {
        if(dp.containsKey(n))
        {
            return dp.get(n);
        }
        else
        {
            dp.put(n,fn(n-1)+fn(n-2));
            return dp.get(n);
        }
    }
    public int fib(int n) {
        dp.put(0,0);
        dp.put(1,1);
        return fn(n);
    }
}